nixos/modules/services/admin/nomachine.nix

```
@@ -0,0 +1,258 @@
+    #CommandXdpyInfo ${pkgs.xorg.xdpyinfo}/bin/xdpyinfo
+    #CommandXmodmap ${pkgs.xorg.xmodmap}/bin/xmodmap
+    #CommandXkbComp ${pkgs.xorg.xkbcomp}/bin/xkbcomp
+    #CommandXauth ${nxserver}/NX/bin/nxauth
+
+    DisplayServerThreads auto
+    DisplayEncoderThreads auto
+    EnableDirectXSupport 0
+  '';
+in
+{
+  options.services.nxserver = {
+    enable = mkEnableOption "the nxserver service";
+
+    enableDebug = mkEnableOption "debug output";
+    debugLevel = mkOption {
+      default = 6;
+      example = 6;
+      description = "Debug output level";
+      type = lib.types.int;
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to open port 4000 in the firewall.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.users = {
+      "nx" = {
+        group = "nx";
+        isSystemUser = true;
+        shell = "${nxserver}/NX/etc/nxserver";
+        home = "/var/lib/nxserver/nx";
+      };
+    };
+
+    users.groups = {
+      "nx" = { };
+    };
+
+    security.pam.services."nx".text = ''
+      # This is a default PAM configuration for NoMachine. It is based on
+      # system's 'su' configuration and can be adjusted freely according
+      # to administrative needs on the system.
+
+      auth       include       su
+      account    include       su
+      password   include       su
+      session    include       su
+    '';
+
+    security.pam.services."nxlimits".text = ''
+      # This is a default PAM configuration for NoMachine.
+      # Used to obtain nx user and nxhtd user limits.
+
+        session   optional pam_limits.so
+    '';
+
+    environment.etc = {
+      "NX/server/localhost/server.cfg" = {
+        text = ''ServerRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/server/localhost/node.cfg" = {
+        text = ''NodeRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/server/localhost/client.cfg" = {
+        text = ''ClientRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/nxserver".source = "/run/wrappers/bin/nxserver";
+      "NX/nxnode".source   = "${nxserver}/bin/nxnode";
+
+      "NX/server.cfg".text = serverConf;
+      "NX/node.cfg".text = nodeConf;
+    };
+
+    security.wrappers = {
+      nxexec = {
+        source = "${nxserver}/NX/bin/nxexec.orig";
+        owner = "root";
+        group = "root";
+        setuid = true;
+      };
+
+      nxserver = {
+        source = "${nxserver}/bin/nxserver";
+        owner = "root";
+        group = "root";
+        setuid = true;
+        setgid = true;
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) [ 4000 ];
+
+    systemd.services."nxserver" = {
+      description = "NoMachine Server daemon";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "syslog.target" "network.target" "network-online.target" ];
+      wants = [ "network-online.target" ];
```

`nxserver` doesn't seem to like it when the display manager is started before it or restarted while it's running. I had to manually restart it a lot while testing, but to make this automatic, we could bind it to `display-manager.service` and make sure it starts after it.
```suggestion
      after = [ "syslog.target" "network.target" "network-online.target" "display-manager.service" ];
      wants = [ "network-online.target" ];
      bindsTo = [ "display-manager.service" ];
```

---

I blindly copied what you suggested, I had no issue so far. Could you tell me how to reproduce the issue so I can test is as well?